### DISCUSSION TITLE
my png get blurred when using in sprite. want to know best practice to make PNG.

### DISCUSSION THREAD

**Discussion by humbleCoder1990** [pixijs/pixijs#7636] [Q&A]
Hi Pixi team.
I have a question about right way to make a png to be used as Pixi.Texture.
I made a pad image on Adobe PhotoShop (500 x 500)

![image](https://user-images.githubusercontent.com/87828192/126689217-e8459b91-f6c1-4e7f-ad43-686a89a9427e.png)

Then, I put it as texture, then used in multiple sprites.
Those sprites's width & height are around (30 x 30) 
Then it get blurred like this.
![image](https://user-images.githubusercontent.com/87828192/126689537-63f5081c-84d5-4f96-af49-7c3dbc903a5b.png)

The reason why I made 500 x 500 png is to prepare large resolution display.
It seems downscaling via sprite's width & height (not sprite.scale) makes my png get blurred.
Should I make a fixed multiple pngs to prepare various display sizes?

MyPad_500_500.png
MyPad_100_100.png
MyPad_50_50.png

like this?

I can't guess my exact sprite size, because it's caculated via window.innerWidth & window.innerHeight at first open.

Sorry for my bad english!
Thanks

**Comment by bigtimebuddy** [comment]
There are a few ways I'd answer this question.

## Scale Mode

The "blur" is being seen because the default in Pixi for images to set `scaleMode` to be linear (`SCALE_MODES.LINEAR`). By changing this value to nearest (`SCALE_MODES.NEAREST`), you can remove a lot of the blurring, but will probably replace it with a pixelated look. Which is probably not great either.

# Resolution

Generally, you want to load to image size that is 100% scaled, then you know exactly how it's going to look. It's also better to reduce the amount of memory you'll need for texture to only load what is necessary. May developers use screen resolution to figure out what size is needed and use a naming convention like this:

* `MyPad-@2x.png` (2x resolution, e.g. retina screens)
* `MyPad-@1x.png` (1x resolution)

But it's your responsibility as the developer to figure out which size you need to load on what device.

## Generate the Texture

The other option is that you can generate this texture on-the-fly. So you could use Graphics plus a bevel effect like [BevelFilter](https://github.com/pixijs/filters) to achieve the look, then you render this to a RenderTexture and use it like you would any other Texture.

**Reply by humbleCoder1990** [reply]
Thanks Matt for an amazing tutorial.
I'm a alone developer who lives in Korea. You are the sunshine.

I have a few questions.

1. If I have 1000 different png files loaded on Pixi.Texture (total file sizes :  10kb * 1000),
will my app's memory usage also be increased like + 10kb * 1000 ?

2. Will a Texture made by Graphics + BevelFilter use less memory than using exactly same looks png file from Photostop?

3. Should I worry about memory as using png? What's the max you recommend? Should I try to use less png files ?

4. If I generate texture using Graphics, Can I use it direclty without waiting kind of "Loading-time"?
Because I saw some codes about "app.load" to wait all assets are loaded.

5. If I try to anmate BevelFilter's parameters on run-time (for example, inside app.ticker()) , 
Is it expenssive (a lot of cpu usages)? or is it cheap like tint?

Thanks Matt again!

**Reply by humbleCoder1990** [reply]
![image](https://user-images.githubusercontent.com/87828192/126704967-39de977b-a75a-45ec-9a0a-f5f9c4d4860c.png)

Sorry to post here about pixi-filters.
But it gives me this error to use it in Typescript.
But I can't install @types/pixi-filters

**Reply by miltoncandelero** [reply]
which version of `pixi-filters` do you have inside your `package.json`? 🤔

**Reply by humbleCoder1990** [reply]
![image](https://user-images.githubusercontent.com/87828192/126708821-a2cdc83c-ce08-48b8-bbb9-65d2a0534314.png)
I just installed it today.
I'm using Vite.

**Reply by humbleCoder1990** [reply]
I should export filter like this

import { BevelFilter } from '@pixi/filter-bevel'

I hope 'pixi-filters' works instead @pixi

**Reply by bigtimebuddy** [reply]
`pixi-filters` missing types is a bug, I'll fix that.

**Reply by humbleCoder1990** [reply]
![Kapture 2021-07-23 at 06 32 21](https://user-images.githubusercontent.com/87828192/126712240-3ce6c762-59e3-4d26-9fcb-d32add9a6573.gif)

I tried my own test with animating filter parameters on app.ticker().add()
It seems very expenssive.
I have 64 sprites with tint, and they have own BevelFilter.
I animated 'thickness' option. CPU almost hit around 70-80%.

**Reply by ivanpopelyshev** [reply]
pixi doesnt have cheap bevel for sprites. You can make a mesh-shader https://pixijs.io/examples/#/mesh-and-shaders/shadertoy-
mesh.js  , https://pixijs.io/examples/#/mesh-and-shaders/triangle-textured.js

why it wasnt done before? it was. in private repos.
why not to put it in pixi? it still requires work on simplifying, and some people just dont want to spend time collaborating.
please, if you make shader like that - share it as a pixi plugin.
